src/opt/horizontal_fusion_impl.cpp

```diff
@@ -0,0 +1,839 @@
+            return false;
+    }
+
+    return true;
+}
+
+void horizontal_fusion_impl::concat(std::vector<instruction_ref>& instrs,
+                                    std::unordered_map<instruction_ref, instruction_ref>& root,
+                                    int root_axis)
+{
+    instruction_ref ins0               = instrs.at(0);
+    instruction_ref base               = root[ins0];
+    std::vector<std::size_t> base_lens = base->get_shape().lens();
+    int axis                           = find_axis(ins0, base, root_axis);
+
+    int sum      = 0;
+    int base_sum = 0;
+    for(auto&& ins : instrs)
+    {
+        sum += ins->get_shape().lens().at(axis);
+        base_sum += root[ins]->get_shape().lens().at(root_axis);
+    }
+
+    base_lens[root_axis] = base_sum;
+
+    if(ins0->outputs().size() != 1)
+        MIGRAPHX_THROW("unexpected output size");
+    instruction_ref output            = ins0->outputs().at(0);
+    shape s                           = ins0->get_shape();
+    std::vector<std::size_t> new_lens = s.lens();
+    new_lens[axis]                    = sum;
+
+    if(ins0->name() == "@literal")
+    {
+        // concat literals.
+        unsigned long long unit_slice   = 1;
+        int ndx                         = 0;
+        unsigned long long new_elements = 1;
+
+        for(auto&& len : s.lens())
+        {
+            if(ndx > axis)
+                unit_slice *= len;
+            if(ndx != axis)
+                new_elements *= len;
+            ndx++;
+        }
+        new_elements *= sum;
+        unsigned type_size             = s.type_size();
+        unsigned long long total_bytes = new_elements * type_size;
+        std::shared_ptr<char> input    = make_shared_array<char>(total_bytes);
+        std::vector<unsigned long long> bytes_per_slice;
+        unsigned long long unit_slice_bytes = unit_slice * type_size;
+
+        std::transform(instrs.begin(),
+                       instrs.end(),
+                       std::back_inserter(bytes_per_slice),
+                       [&](auto&& d) -> unsigned long long {
+                           return d->get_shape().lens().at(axis) * unit_slice_bytes;
+                       });
+
+        unsigned copy_bytes = 0;
+        int slice_ndx       = 0;
+        char* dst           = input.get();
+        while(copy_bytes < total_bytes)
+        {
+            unsigned ins_ndx = 0;
+            for(auto&& ins : instrs)
+            {
+                unsigned long long bytes = bytes_per_slice[ins_ndx];
+                const char* src          = &(ins->get_literal().data()[slice_ndx * bytes]);
+                std::copy(src, src + bytes, dst);
+                dst += bytes;
+                copy_bytes += bytes;
+                ins_ndx++;
+            }
+            slice_ndx++;
+        }
+        shape new_shape{s.type(), new_lens};
+        auto new_literal = p_program->add_literal(literal(new_shape, input.get()));
```

Comment:
Can we just insert a concat instruction instead? The constant propogation will take care of replacing the literal. There is no need to implement concat twice.

Reply:
Tried insert concat, it doesn't work.  propagate_constant only removes concat whose axis is 0.

Reply:
This should be fixed in the propagate_constant pass and not in horizontal fusion. Can you show me the test case that is failing? For me this test case passes when using axis=1 for concat:

```cpp
TEST_CASE(const_concat)
{
    migraphx::program p1;
    {
        auto lit = migraphx::literal{{migraphx::shape::float_type, {2, 1, 2}}, {1, 2, 3, 4}};
        auto x1 = p1.add_literal(lit);
        auto x2 = p1.add_literal(lit);
        auto concat = p1.add_instruction(migraphx::op::concat{1}, x1, x2);
        p1.add_instruction(pass_op{}, concat);
    }
    p1.compile(const_prop_target{});

    migraphx::program p2;
    {
        auto lit = migraphx::literal{{migraphx::shape::float_type, {2, 2, 2}}, {1, 2, 1, 2, 3, 4, 3, 4}};
        auto x = p2.add_literal(lit);
        p2.add_instruction(pass_op{}, x);
    }
    EXPECT(p1 == p2);
}
```

Reply:
Ah wait, I see the issue. The horizontal fusion should be moved before the propagate_constant pass.

Reply:
I just added a propagate_constant  and a dead store elimination after horizontal fusion.
And used concat instead.  So I deemed this one fixed.